From 0951d99a31bbb29f739278ff59ebe5d82655f5c8 Mon Sep 17 00:00:00 2001
From: Wolfgang Grandegger <wg@aries-embedded.de>
Date: Fri, 28 Apr 2023 15:47:55 +0200
Subject: [PATCH 8/8] board: aries: mcvevk: support the Renesas 5P35023 clock
 generator

The new MCV modules with hw rev. 4.1 use the Renesas 5P35023 clock
generator, which needs to be initialized at startup via I2C bus at
address 0x68. Old modules are still supported. No I2C device will
be found at that address.

Because the clock starts late, the phy address strapping does not
work properly. After power-on, the PHY is found at a different MII
address than 0x0. If we press the reset botton on the MCVEVP,
which resets the PHY, it shows up at the expected adress 0x0.
To work around that issue, we initialize the clock chip once in
the SPL and perform a warm reset to make the phy work properly.

Signed-off-by: Wolfgang Grandegger <wg@aries-embedded.de>
---
 .../dts/socfpga_cyclone5_mcvevk-u-boot.dtsi   |  8 ++
 board/aries/mcvevk/Makefile                   |  4 +
 board/aries/mcvevk/spl.c                      | 73 +++++++++++++++++++
 configs/socfpga_mcvevk_defconfig              |  2 +
 4 files changed, 87 insertions(+)
 create mode 100644 board/aries/mcvevk/spl.c

diff --git a/arch/arm/dts/socfpga_cyclone5_mcvevk-u-boot.dtsi b/arch/arm/dts/socfpga_cyclone5_mcvevk-u-boot.dtsi
index eea453b8ad..4680337b05 100644
--- a/arch/arm/dts/socfpga_cyclone5_mcvevk-u-boot.dtsi
+++ b/arch/arm/dts/socfpga_cyclone5_mcvevk-u-boot.dtsi
@@ -12,6 +12,10 @@
 	status = "disabled";
 };
 
+&i2c0 {
+	u-boot,dm-pre-reloc;
+};
+
 &mmc {
 	u-boot,dm-pre-reloc;
 };
@@ -32,3 +36,7 @@
 &portc {
 	bank-name = "portc";
 };
+
+&rst {
+	u-boot,dm-pre-reloc;
+};
diff --git a/board/aries/mcvevk/Makefile b/board/aries/mcvevk/Makefile
index e1c8a6b3c7..964bcc8c20 100644
--- a/board/aries/mcvevk/Makefile
+++ b/board/aries/mcvevk/Makefile
@@ -5,3 +5,7 @@
 # (C) Copyright 2010, Thomas Chou <thomas@wytron.com.tw>
 
 obj-y	:= socfpga.o
+
+ifdef CONFIG_SPL_BUILD
+obj-y += spl.o
+endif
diff --git a/board/aries/mcvevk/spl.c b/board/aries/mcvevk/spl.c
new file mode 100644
index 0000000000..2ac0463082
--- /dev/null
+++ b/board/aries/mcvevk/spl.c
@@ -0,0 +1,73 @@
+#include <common.h>
+#include <spl.h>
+#include <i2c.h>
+#include <asm/io.h>
+#include <asm/arch/reset_manager.h>
+
+DECLARE_GLOBAL_DATA_PTR;
+
+#define MCVEVK_SP35023_I2C_BUS	0
+#define MCVEVK_SP35023_I2C_ADDR 0x68
+
+struct sp35023_regs {
+	unsigned char addr;
+	unsigned char val;
+};
+
+
+#define MCVEVK_SP35023_CLK_ENABLE_REG 0x23
+#define MCVEVK_SP35023_CLK_ENABLE_VAL 0x41
+static struct sp35023_regs mcvevk_sp35023_regs[] = {
+	{0x1b,0x37},
+	{0x21,0xc0},
+	{0x24,0x8f},
+	{0x1f,0xc7},
+	{0x00,0x08},
+	{0x0f,0x00},
+	{MCVEVK_SP35023_CLK_ENABLE_REG, MCVEVK_SP35023_CLK_ENABLE_VAL},
+};
+
+void spl_board_init(void)
+{
+	struct udevice *sp35023 = NULL;
+	int i, err;
+	uint8_t val;
+	volatile unsigned int *reset_mgr_ctrl =
+		(volatile unsigned int *)0xffd05004;
+
+	/*
+	 * Configure the Renesas 5P35023 Clock Generator available
+	 * on hw revision >= 4.1
+	 */
+	err = i2c_get_chip_for_busnum(MCVEVK_SP35023_I2C_BUS,
+				      MCVEVK_SP35023_I2C_ADDR, 1, &sp35023);
+	if (err)
+		return;
+
+	err = dm_i2c_read(sp35023, MCVEVK_SP35023_CLK_ENABLE_REG,
+			  &val, sizeof(val));
+	if (err) {
+		printf("Couldn't read sp35023 clock enable reg\n");
+		return;
+	}
+	if (val == MCVEVK_SP35023_CLK_ENABLE_VAL) {
+		return;	/* Clock is already running */
+	}
+
+
+	/* Initialize and start clock generator */
+	puts("Initializing PLL...\n");
+	for (i = 0; i < ARRAY_SIZE(mcvevk_sp35023_regs); i++) {
+		unsigned char addr = mcvevk_sp35023_regs[i].addr;
+		unsigned char val = mcvevk_sp35023_regs[i].val;
+
+		err = dm_i2c_write(sp35023, addr, &val, sizeof(val));
+		if (err) {
+			printf("Couldn't write %#x to sp35023 reg at %#x\n",
+			       val, addr);
+			return;
+		}
+	}
+	puts("Warm reset...\n");
+	*reset_mgr_ctrl = BIT(RSTMGR_CTRL_SWWARMRSTREQ_LSB);
+}
diff --git a/configs/socfpga_mcvevk_defconfig b/configs/socfpga_mcvevk_defconfig
index 3c76cfa9c3..25705aa38c 100644
--- a/configs/socfpga_mcvevk_defconfig
+++ b/configs/socfpga_mcvevk_defconfig
@@ -19,6 +19,8 @@ CONFIG_SYS_CONSOLE_OVERWRITE_ROUTINE=y
 CONFIG_SYS_CONSOLE_ENV_OVERWRITE=y
 # CONFIG_DISPLAY_BOARDINFO is not set
 CONFIG_DISPLAY_BOARDINFO_LATE=y
+CONFIG_SPL_BOARD_INIT=y
+CONFIG_SPL_I2C=y
 CONFIG_SPL_SPI_LOAD=y
 CONFIG_SYS_SPI_U_BOOT_OFFS=0x40000
 CONFIG_CMD_ASKENV=y
-- 
2.34.1

